pkg/observer/observer.go

```diff
@@ -0,0 +1,218 @@
+
+	// Put puts the DCAS value and returns the key for the value
+	Put(ns, coll string, value []byte) (string, error)
+}
+
+type cfg interface {
+	GetChannels() []string
+}
+
+// Start starts observer routines
+func Start(cfg cfg) error {
+
+	if roles.HasRole(observerRole) {
+
+		logger.Infof("peer is an observer, channels to observe: %s", cfg.GetChannels())
+		for _, channelID := range cfg.GetChannels() {
+			initObservingOnChannel(channelID)
+		}
+
+	} else {
+		logger.Info("peer is not an observer, nothing to do...")
+	}
+
+	return nil
+}
+
+func initObservingOnChannel(channelID string) {
+
+	logger.Debugf("initialize observing on channel: %s", channelID)
+
+	// register to receive Sidetree transactions from blocks
+	n := notifier.New(getBlockPublisher(channelID))
+
+	sidetreeTxnChannel := n.RegisterForSidetreeTxn()
+
+	go func(channel string, txnsCh <-chan notifier.SidetreeTxn) {
+
+		for {
+			txn, ok := <-txnsCh
+			if !ok {
+				logger.Warnf("received !ok from channel '%s'", channel)
+				return
+			}
+
+			err := processSidetreeTxn(channel, txn)
+			if err != nil {
+				logger.Warnf("Failed to process anchor[%s] on channel[%s]: %s", txn.AnchorAddress, channel, err.Error())
+				continue
+			}
+
+			logger.Debugf("Successfully processed anchor[%s] on channel[%s]", txn.AnchorAddress, channel)
+
+		}
+	}(channelID, sidetreeTxnChannel)
+
+}
+
+func processSidetreeTxn(channel string, sidetreeTxn notifier.SidetreeTxn) error {
+
+	logger.Debugf("processing sidetree txn:%+v on channel '%s'", sidetreeTxn, channel)
+
+	content, err := getContent(channel, sidetreeTxn.AnchorAddress)
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve content for anchor: channel[%s] sidetreeNs[%s] key[%s]", channel, sidetreeNs, sidetreeTxn.AnchorAddress)
+	}
+
+	logger.Debugf("cas content for anchor[%s]: %s", sidetreeTxn.AnchorAddress, string(content))
+
+	af, err := getAnchorFile(content)
+	if err != nil {
+		return errors.Wrapf(err, "failed to unmarshal anchor[%s]", sidetreeTxn.AnchorAddress)
+	}
+
+	return processBatchFile(channel, af.BatchFileHash)
+}
+
+func processBatchFile(channel string, batchFileAddress string) error {
+
+	content, err := getContent(channel, batchFileAddress)
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve content for batch: channel[%s] sidetreeNs[%s] key[%s]", channel, sidetreeNs, batchFileAddress)
+	}
+
+	bf, err := getBatchFile(content)
+	if err != nil {
+		return errors.Wrapf(err, "failed to unmarshal batch[%s]", batchFileAddress)
+	}
+
+	logger.Debugf("batch file operations: %s", bf.Operations)
+
+	for _, op := range bf.Operations {
+		addr, err := putContent(channel, []byte(op))
+		if err != nil {
+			return errors.Wrapf(err, "failed to store operation[%s] from batch[%s]", op, batchFileAddress)
+		}
+
+		logger.Debugf("successfully stored operation[%s] at address[%s]", op, addr)
+	}
+
+	return nil
+}
+
+var getDCAS = func(channelID string) dcasClient {
+	return dcasclient.New(channelID)
```

Comment:
It's expensive to create a new client every time. The client maintains a cache of collection configs. The first time it's used, it needs to call the ledger to retrieve the collection config. It's better to reuse the client.

Reply:
Done